contrib/win32/win32compat/misc.c

```diff
@@ -1466,3 +1466,77 @@ is_absolute_path(char *path)
 	
 	return retVal;
 }
+
+/* return 0 - in case of failure */
+int
+create_directory_withsddl(char *path, char *sddl)
+{
+	struct stat st;
+	if (stat(path, &st) < 0) {
+		PSECURITY_DESCRIPTOR pSD = NULL;
+		SECURITY_ATTRIBUTES sa;
+		memset(&sa, 0, sizeof(SECURITY_ATTRIBUTES));
+		sa.nLength = sizeof(SECURITY_ATTRIBUTES);
+		sa.bInheritHandle = FALSE;
+
+		wchar_t *path_w = utf8_to_utf16(path);
+		if (!path_w) {
+			fatal("%s utf8_to_utf16() has failed to convert string:%s", __func__, path);
+			return 0; /* to avoid the compiler warning */
+		}
+
+		wchar_t *sddl_w = utf8_to_utf16(sddl);
+		if (!sddl_w) {
+			fatal("%s utf8_to_utf16() has failed to convert string:%s", __func__, sddl);
+			return 0; /* to avoid the compiler warning */
+		}
+
+		if (ConvertStringSecurityDescriptorToSecurityDescriptorW(sddl_w, SDDL_REVISION, &pSD, NULL) == FALSE) {
+			error("ConvertStringSecurityDescriptorToSecurityDescriptorW failed with error code %d", GetLastError());
+			return 0;
+		}
+
+		if (IsValidSecurityDescriptor(pSD) == FALSE) {
+			error("IsValidSecurityDescriptor return FALSE");
+			return 0;
+		}
+
+		sa.lpSecurityDescriptor = pSD;
+		if (!CreateDirectoryW(path_w, &sa)) {
+			error("Failed to create directory:%ls error:%d", path_w, GetLastError());
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+/* return 0 - in case of failure */
+int
+copy_file(char *source, char *destination)
+{
+	if (!source || !destination) return 1;
+
+	struct stat st;
+	if ((stat(destination, &st) < 0) || (stat(source, &st) < 0)) {
+		wchar_t *source_w = utf8_to_utf16(source);
+		if (!source_w) {
+			fatal("%s utf8_to_utf16() has failed to convert string:%s", __func__, source_w);
```

Comment:
is error or returning 0 enough because anyway, the caller call fatal if the returning is 0.